File: hello-tts-rust/src/tts_client.rs
use crate::audio_player::AudioPlayer;
use crate::backends::edge::EdgeTTS;
use crate::backends::google::GoogleTTS;
use crate::backends::TTSBackend;
use crate::config::TTSConfig;
use crate::models::{TTSError, Voice};
use log::{info};
use std::path::Path;
use tokio::fs;

/// TTS Processor that delegates to a configured backend
pub struct TTSProcessor {
    config: TTSConfig,
    voices_cache: Option<Vec<Voice>>,
    backend: Box<dyn TTSBackend + Send + Sync>,
}

impl TTSProcessor {
    /// Create a new TTSProcessor with optional configuration
    pub fn new(config: Option<TTSConfig>) -> Self {
        let config = config.unwrap_or_default();
        let backend: Box<dyn TTSBackend + Send + Sync> = match config.backend.as_str() {
            "google" => Box::new(GoogleTTS::new()),
            _ => Box::new(EdgeTTS::new()),
        };

        Self {
            config,
            voices_cache: None,
            backend,
        }
    }

    /// Convert text to audio data using the configured backend
    pub async fn synthesize_text(&self, text: &str, voice: &str) -> Result<Vec<u8>, TTSError> {
        self.backend.synthesize_text(text, voice).await
    }

    /// Get all available voices from the configured backend
    pub async fn list_voices(&mut self) -> Result<Vec<Voice>, TTSError> {
        if self.config.cache_voices {
            if let Some(ref voices) = self.voices_cache {
                info!("Using cached voices");
                return Ok(voices.clone());
            }
        }

        info!("Fetching voices from backend");
        let voices = self.backend.list_voices().await?;

        if self.config.cache_voices {
            self.voices_cache = Some(voices.clone());
        }

        Ok(voices)
    }

    /// Synthesize text and play it, optionally saving to a file
    pub async fn synthesize_and_play(
        &self,
        text: &str,
        voice: &str,
        output_path: Option<&Path>,
        play: bool,
    ) -> Result<(), TTSError> {
        let audio_data = self.synthesize_text(text, voice).await?;

        if let Some(path) = output_path {
            if let Some(parent) = path.parent() {
                fs::create_dir_all(parent).await?;
            }
            fs::write(path, &audio_data).await?;
        }

        if play {
            let player = AudioPlayer::new().map_err(|e| TTSError::Synthesis(e.to_string()))?;
            player
                .play_audio_data(audio_data, Some("mp3"))
                .map_err(|e| TTSError::Synthesis(e.to_string()))?;
        }

        Ok(())
    }

    /// Clear the cached voice list
    pub fn clear_voice_cache(&mut self) {
        self.voices_cache = None;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_voice_creation() {
        let voice = Voice::new(
            "en-US-AriaNeural".to_string(),
            "Aria".to_string(),
            "en-US".to_string(),
            "Female".to_string(),
        );

        assert_eq!(voice.name, "en-US-AriaNeural");
        assert_eq!(voice.display_name, "Aria");
        assert_eq!(voice.locale, "en-US");
        assert_eq!(voice.gender, "Female");
        assert_eq!(voice.language_code(), "en");
    }

    #[test]
    fn test_voice_matches_language() {
        let voice = Voice::new(
            "en-US-AriaNeural".to_string(),
            "Aria".to_string(),
            "en-US".to_string(),
            "Female".to_string(),
        );

        assert!(voice.matches_language("en"));
        assert!(voice.matches_language("en-US"));
        assert!(!voice.matches_language("fr"));
    }

    #[test]
    fn test_tts_config_default() {
        let config = TTSConfig::default();
        assert_eq!(config.default_voice, "en-US-AriaNeural");
        assert_eq!(config.output_format, "mp3");
        assert_eq!(config.max_retries, 3);
    }

    #[tokio::test]
    async fn test_tts_client_creation() {
        let client = TTSProcessor::new(None);
        assert_eq!(client.config.default_voice, "en-US-AriaNeural");
    }
}

impl TTSProcessor {
    pub async fn save_audio(&self, audio_data: &[u8], filename: &str) -> Result<(), TTSError> {
        self.backend.save_audio(audio_data, filename).await
    }
}

File: hello-tts-python/backends/edge/tts_client.py
"""Edge TTS client implementation."""
from config_manager import TTSConfig
from voice import Voice
import logging
import edge_tts
import aiofiles
import os
from typing import List, Optional
import sys
from pathlib import Path

# Add parent directories to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent.parent))


logger = logging.getLogger(__name__)


class TTSError(Exception):
    pass


class TTSClient:
    def __init__(self, config: Optional[TTSConfig] = None):
        self.config = config or TTSConfig()
        self._voices_cache: Optional[List[Voice]] = None

    async def synthesize_text(self, text: str, voice: str) -> bytes:
        """Synthesize text to audio using Edge TTS."""
        communicate = edge_tts.Communicate(text, voice)
        audio_data = b""
        async for chunk in communicate.stream():
            if chunk["type"] == "audio":
                audio_data += chunk["data"]

        if not audio_data:
            raise TTSError("No audio data generated")

        return audio_data

    async def save_audio(self, audio_data: bytes, filename: str) -> None:
        async with aiofiles.open(filename, 'wb') as f:
            await f.write(audio_data)

    async def list_voices(self) -> List[Voice]:
        if self._voices_cache is None:
            # Try to load from config file first
            config_path = "shared/tts_config.json"
            if os.path.exists(config_path):
                try:
                    self._voices_cache = Voice.parse_voices_from_json_file(
                        config_path)
                    return self._voices_cache
                except Exception as e:
                    logger.warning(
                        f"Failed to load voices from config file: {e}")

            # Fallback to API
            voices_data = await edge_tts.list_voices()
            self._voices_cache = [
                Voice(
                    name=v["ShortName"],
                    display_name=v.get("FriendlyName", v.get("ShortName", "")),
                    locale=v.get("Locale", ""),
                    gender=v.get("Gender", "unknown")
                )
                for v in voices_data
            ]
        return self._voices_cache

    async def get_voices_by_language(self, language: str) -> List[Voice]:
        all_voices = await self.list_voices()
        return [v for v in all_voices if v.locale.startswith(language) or v.locale.split('-')[0] == language]

    def clear_voice_cache(self) -> None:
        self._voices_cache = None
